source/client-side-encryption/client-side-encryption.rst

```diff
@@ -2136,6 +2204,15 @@ command with maxMessageSizeBytes=24MB. If after marking we determine
 that's too large, try again with maxMessageSizeBytes=12MB and so on. And
 in the end libmongocrypt would create multiple OP_MSGs to send.
 
+Support sessions in key management functions
+--------------------------------------------
+
+Key management functions currently assume there are no concurrent accesses of
+the key vault collection being operated on. To support concurrent access of the
+key vault collection, the key management functions may be overloaded to take an
+explicit session parameter as described in the
+`Drivers Sessions Specification <../sessions/driver-sessions.rst>`__.
```

Comment:
“explicit session parameter” sounds like the signature would be required to be e.g. `rewrapManyDataKey(filter, session, opts)` or `rewrapManyDataKey(filter, opts, session)`. At least in the case of Node.js, it would be significantly easier (and match existing driver API usage) to make `session` a possible entry in `opts`. Would that be considered acceptable?

Reply:
This wording is borrowing the specification in `driver-sessions.rst` as linked:

>  All ``MongoDatabase`` methods that talk to the server SHOULD be overloaded to take an explicit session parameter. (See `why is session an explicit parameter?`_.)

I do not know if it would be considered acceptable to use a field in `opts` as an alternative. CC @jmikola.

Reply:
The relevant text in [that spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#new-database-methods-that-take-an-explicit-session) is:

> When overloading methods to take a session parameter, the session parameter SHOULD be the first parameter. If overloading is not possible for your language, it MAY be in a different position or MAY be embedded in an options structure.

A number of drivers just take a session through an options struct (or keyword args) instead of overloading.

With respect to unified spec tests, the session is parameter is merely expressed as the `session` key in the `arguments` object of an [operation](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#operation).

Reply:
Cool – I think that clarifies everything I need to know here (i.e. yes, in the Node.js driver it should probably be part of the options object). Feel free to resolve :+1: